rust/lance/src/index/vector/hnsw.rs

```diff
@@ -141,24 +136,41 @@ impl Index for HNSWIndex {
 #[async_trait]
 impl VectorIndex for HNSWIndex {
     #[instrument(level = "debug", skip_all, name = "HNSWIndex::search")]
-    async fn search(&self, query: &Query, _pre_filter: Arc<PreFilter>) -> Result<RecordBatch> {
+    async fn search(&self, query: &Query, pre_filter: Arc<PreFilter>) -> Result<RecordBatch> {
+        let row_ids = self.hnsw.storage().row_ids();
+        let bitmap = if pre_filter.is_empty() {
+            None
+        } else {
+            pre_filter.wait_for_ready().await?;
+
+            let indices = pre_filter.filter_row_ids(row_ids);
+            Some(
+                RoaringBitmap::from_sorted_iter(indices.into_iter().map(|i| i as u32)).map_err(
+                    |e| Error::Index {
+                        message: format!("Error creating RoaringBitmap: {}", e),
+                        location: location!(),
+                    },
+                )?,
+            )
+        };
+
         let results = self.hnsw.search(
             query.key.as_primitive::<Float32Type>().as_slice(),
             query.k,
             30,
-            None,
+            bitmap,
         )?;
 
-        let node_ids = UInt32Array::from_iter_values(results.iter().map(|x| x.0));
+        let row_ids = UInt64Array::from_iter_values(results.iter().map(|x| row_ids[x.0 as usize]));
```

Comment:
So `search` returns indices into the storage's row ids?

Reply:
yes, here the result is `(node_id, dist)`, and `node_id` is the indice of row